R/zchunk_L244.building_det.R

```diff
@@ -0,0 +1,1044 @@
+    L244.DemandFunction_flsp <- write_to_all_regions(A44.demandFn_flsp, LEVEL2_DATA_NAMES[["DemandFunction_flsp"]],
+                                                     GCAM_region_names = GCAM_region_names)
+
+    # Floorspace demand satiation
+    # L244.Satiation_flsp: Satiation levels assumed for floorspace
+    L244.Satiation_flsp_class <- A44.satiation_flsp %>%
+      gather(sector, value, resid, comm) %>%
+      mutate(satiation.level = value * CONV_THOUS_BIL) %>%
+      select(-value)
+
+    L244.Satiation_flsp <- write_to_all_regions(A44.gcam_consumer, c("region", "gcam.consumer", "nodeInput", "building.node.input"), # replace with LEVEL2_DATA_NAMES[["BldNodes]]
+                                                GCAM_region_names = GCAM_region_names) %>%
+      # Match in the region class, and use this to then match in the satiation floorspace
+      left_join_error_no_match(A_regions %>% select(region, region.class),
+                               by = "region") %>%
+      left_join_error_no_match(L244.Satiation_flsp_class, by = c("region.class", "gcam.consumer" = "sector")) %>%
+      select(LEVEL2_DATA_NAMES[["Satiation_flsp"]])
+
+    # Satiation adder - Required for shaping the future floorspace growth trajectories in each region
+    # L244.SatiationAdder: Satiation adders in floorspace demand function
+    # First, prepare socioeconomics tables by adding region names
+    L102.pcgdp_thous90USD_Scen_R_Y <- L102.pcgdp_thous90USD_Scen_R_Y %>%
+      left_join_error_no_match(GCAM_region_names, by = "GCAM_region_ID") %>%
+      rename(pcGDP_thous90USD = value)
+    L101.Pop_thous_R_Yh <- L101.Pop_thous_R_Yh %>%
+      left_join_error_no_match(GCAM_region_names, by = "GCAM_region_ID") %>%
+      rename(pop_thous = value)
+
+    # In order to pass timeshift, we need floorspace for energy satiation year, which we don't have
+    # So instead, we will take floorspace for maximum year
+    Floorspace_timeshift_pass <- L244.Floorspace %>%
+      filter(year == max(L244.Floorspace$year)) %>%
+      select(-year)
+
+    # Match in the per-capita GDP, total floorspace, and population (for calculating per-capita floorspace)
+    L244.SatiationAdder <- L244.Satiation_flsp %>%
+      left_join_keep_first_only(L102.pcgdp_thous90USD_Scen_R_Y %>%
+                                  filter(year == energy.SATIATION_YEAR), by = "region") %>%
+      # timeshift error here
+      left_join_error_no_match(Floorspace_timeshift_pass, by = c("region", "gcam.consumer", "nodeInput", "building.node.input")) %>%
+      left_join_error_no_match(L101.Pop_thous_R_Yh, by = c("region", "year", "GCAM_region_ID")) %>%
+      mutate(pcFlsp_mm2 = base.building.size / pop_thous,
+             # We now have all of the data required for calculating the satiation adder in each region
+             satiation.adder = round(satiation.level - exp(log(2) * pcGDP_thous90USD / energy.GDP_MID_SATIATION) *
```

Comment:
An explanatory comment would probably be good here--e.g. what's the `log` calculating?

Reply:
Yeah, any sort of information about what is going on here would be helpful.

Reply:
See my comment above--this is just reverse engineering the building satiation equation in the C++ code so that the satiation function (i.e., the fraction of the assumed satiation level as a function of per-capita GDP in any time period) passes through the calibration year's data point, and will hit 50% of the assumed satiation level at energy.GDP_MID_SATIATION (currently $10,500 per capita in 1990$). Obvi this could have been easier if we just read in the energy.GDP_MID_SATIATION instead!